imnodes.cpp

```diff
@@ -1679,6 +1693,24 @@ void draw_pin(EditorContext& editor, const int pin_idx, const bool left_mouse_cl
     draw_pin_shape(pin.pos, pin, pin_color);
 }
 
+// TODO: Separate hover code from drawing code to avoid this unpleasant divergent function signature.
+bool is_node_hovered(EditorContext& editor, const int node_idx)
```

Comment:
We could just pass `const NodeData&` since the only place where it is used already fetches the node

Reply:
One of the things that makes this function confusing (why we need pin data) is the way that the hover code is intertwined with the drawing code. From my experience working on another node editor, the UI should be in reverse of the drawing.

Example: First, we check if we are hovering over a pin. If we are not hovering over a pin, we check if we are hovering over a node. If we are not hovering over a pin or node, we check if we are hovering over a link. Then we draw links, followed by nodes, followed by pins.

The best solution IMO would be to do exactly like we are doing with links right now --- first, draw the pins, then draw the nodes underneath.

The other inconsistency (the need for node_idx) is caused by the the way that node hovering is being calculated outside of  the drawing function. I actually think this is necessary, and perhaps we can do the same with links (and pins, too), since one of the bugs right now is that you can hover over multiple links (and pins, too).

Reply:
I'm going to do a quick fix and come back to this issue later.

Reply:
Right, now I see the bigger picture. It's does seem like, were hovering and rendering implemented the way you described, there would be less complex logic needed. An elegant solution.

Not totally sure why I ended up doing it in this tangled manner originally -- I think I originally made nodes an invisible button to detect hovering and selection, and that resulted in hover detection getting entangled with drawing the elements. I don't think there should be anything anymore preventing one from rendering elements independently from each other, in completely separate passes, in the order you described. There's some complexity with the node rendering order, due to the need to keep them sorted in the same depth order from frame to frame. But that machinery should be quite separate from everything else.

Reply:
Right. Just realized you can’t draw all the pins on top of all the nodes.

Reply:
Right. Just realized you can’t just draw all the pins on top of all the nodes. 😬